Add OrcaBus ID

reisingerf opened the pull request:
A potential option to add a OrcaBus ID as primary key to the model classes. 

Not sure this is a good way... looks convenient though ;-)

williamputraintan reviewed (commented):
The ordering at the viewset level will also need to change to `-orcabus_id`  instead of `-id`

victorskl commented:
... will review in tick

reisingerf commented on lib/workload/stateless/stacks/workflow-manager/workflow_manager/fields.py:
```diff
@@ -1,8 +1,34 @@
+        return name, path, args, kwargs
+
+    def pre_save(self, instance, add):
+        setattr(instance, self.attname, orcabus_id(self.prefix))
```
Ah, yes, it will. 
We wanted to avoid that, right....

williamputraintan replied:
Yup, unless we wanted to enforce this rule depending on each microservice.

victorskl reviewed (approved):
LGTM

reisingerf commented:
What do you guys think: It is still useful to have a `OrcabusIdField`?

alexiswl commented:
Yes, orcabus id field is useful!

reisingerf commented:
@alexiswl as an explicit Django `Field` or simply as part of the model?

For reference, there are two options at the moment:
- the metadata manager is using a `BaseModel` class that handles the Orcabus ID. Other model classes simply extend it to inherit the Orcabus ID.
- in the workflow manager I was experimenting with having an explicit Django custom `Field` (OrcabusIdField`) that could be used instead of another primary key field, so the model classes would still need to specify it

Not sure I can see all the advantages/disadvantage yet...

reisingerf commented:
Digging and playing a bit more, I think Will's approach is better. 
Unless there are strong preferences against it, I'll kill this PR and we'll adopt Will's approach.

victorskl commented:
Go for it, Flo! Either way you feel like. 
This is implementation details. So, we shall respect your choice.

williamputraintan commented:
I like the idea of having `OrcabusIdField` as a custom field so that the generation of ULID is contained within that field. If all models will have OrcabusId, then perhaps we could put this custom Django field in the base model.

williamputraintan commented:
> Do we then need a custom field for a single use in the base class?

Fair enough, I guess the question now is whether we want all models to have the ULID or not. 
Perhaps if we would refer payloads/states in the future then making it a ULID form is worth it? Or to future proof instead, we need to refer any of these objects for a consumer.

> If it's just to contain the ULID, we could possibly simplify the base model by specifying the ID generation mechanism as a field attribute (default=ulid.new) instead of having it in the save() method, but you'd need to overwrite the latter anyway to make sure the validator is called (via full_clean()) right?

True, we could move back the ULID generation to the default (did this way because initially prefix was saved to db, and didn't think creating the custom Django field).  The full_clean just to make sure the data content is according to its type and constraint else trigger an error.

reisingerf commented:
OK, the last changes are to try to align the approach with the Metadata Manager one. 
Not fully tested yet and I still need to expand on the mock data so the API can be used for all use cases, but it should give a good idea.

FYI @williamputraintan @raylrui  this will break the current API ( I removed the `state` and `library` sub-endpoints to workflowrun, as that data is now available in the detailed view of each WorkflowRun record). Hope that makes sense!?

raylrui commented:
>I removed the state and library sub-endpoints to workflowrun, as that data is now available in the detailed view of each WorkflowRun record

yes, all good for me, currently workflowRun page just use:
```
'/api/v1/workflow/'
'/api/v1/workflowrun/'
'/api/v1/workflowrun/{id}/'
'/api/v1/workflowrun/{workflowrunId}/state/'
'/api/v1/payload/{id}/'
'/api/v1/workflowrun/count_by_status/'
```

reisingerf commented:
'/api/v1/workflowrun/{workflowrunId}/state/'

This would disappear. The same info should be in '/api/v1/workflowrun/{workflowrunId}' though.

reisingerf commented:
'/api/v1/workflowrun/count_by_status/'

I don't think I even have that in my branch yet...

raylrui commented:
> '/api/v1/workflowrun/{workflowrunId}/state/'
sure, we can remove it, as long as we can get it. but I still hope we can keep it somewhere, as I am thinking separating workflow run 'detail' with 'state history' view somehow, just thinking...

>'/api/v1/workflowrun/count_by_status/'
this is just recently merge in main branch I think

reisingerf commented:
>I still hope we can keep it somewhere, as I am thinking separating workflow run 'detail' with 'state history' view

Good point!
The state history is now part of the detail view, but we may actually want to change that, so that the WorkflowRun detail view only shows the current/latest state and the state history becomes available via the separate endpoint. 
I think that would make more sense and also adds less payload to the workflow detail view, which otherwise may get quite big...

reisingerf commented:
@raylrui How about this? 

I merged with the changes from `main`, updated the WorkflowRun views and added a few random states to the mock data so the API produces some reasonable responses.

raylrui commented:
@reisingerf Looks Great to me!
